File: Demo/TyphoonRestClientDemo/TyphoonRestClientDemo/ValueTransformers/TRCValueTransformerDateISO8601.m


#import "TRCValueTransformerDateISO8601.h"
#import "TRCUtils.h"

@implementation TRCValueTransformerDateISO8601

+ (NSDateFormatter *)sharedDateFormatter
{
    static NSDateFormatter *dateFormatter = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ssZ"];
        NSLocale *posix = [[NSLocale alloc] initWithLocaleIdentifier:@"en_US_POSIX"];
        [dateFormatter setLocale:posix];
    });
    return dateFormatter;
}

- (NSDate *)objectFromResponseValue:(NSString *)responseValue error:(NSError **)error
{
    NSDateFormatter *dateFormatter = [[self class] sharedDateFormatter];
    
    NSDate *date = [dateFormatter dateFromString:responseValue];
    if (!date && error) {
        *error = NSErrorWithFormat(@"Can't create NSDate from string '%@'", responseValue);
    }
    return date;
}

- (NSString *)requestValueFromObject:(id)object error:(NSError **)error
{
    if (![object isKindOfClass:[NSDate class]]) {
        if (error) {
            *error = NSErrorWithFormat(@"Can't convert '%@' into NSString using %@", [object class], [self class]);
        }
        return nil;
    }
    NSDateFormatter *dateFormatter = [[self class] sharedDateFormatter];
    
    NSString *string = [dateFormatter stringFromDate:object];
    
    if (!string && error) {
        *error = NSErrorWithFormat(@"Can't convert NSDate '%@' into NSStrign", object);
    }
    
    return string;
}

@end

File: Demo/TyphoonRestClientDemoSwift/TyphoonRestClientDemoSwift/TyphoonRestClientDemoSwift/ValueTransformers/TRCValueTransformerDateISO8601.swift
////////////////////////////////////////////////////////////////////////////////
//
//  APPS QUICKLY
//  Copyright 2015 Apps Quickly Pty Ltd
//  All Rights Reserved.
//
//  NOTICE: Prepared by AppsQuick.ly on behalf of Apps Quickly. This software
//  is proprietary information. Unauthorized use is prohibited.
//
////////////////////////////////////////////////////////////////////////////////


import Foundation

class TRCValueTransformerDateISO8601: NSObject, TRCValueTransformer {
    
    static let sharedFormatter = NSDateFormatter()
    
    override class func initialize() {
        sharedFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"
        sharedFormatter.locale = NSLocale(localeIdentifier: "en_US_POSIX")
    }
    
    func objectFromResponseValue(responseValue: AnyObject!, error: NSErrorPointer) -> AnyObject! {
        
        let date = TRCValueTransformerDateISO8601.sharedFormatter.dateFromString(responseValue as! String)
        
        if date == nil && error != nil {
            
            error.memory = NSError(domain: "", code: 0, userInfo: [ NSLocalizedDescriptionKey: "Can't create NSDate from string \(responseValue)"])
        }
        
        return date
    }
    
    
    func requestValueFromObject(object: AnyObject!, error: NSErrorPointer) -> AnyObject! {
        
        if object.isKindOfClass(NSDate.classForCoder()) == false {
            
            if error != nil {
                error.memory = NSError(domain: "", code: 0, userInfo: [ NSLocalizedDescriptionKey: "input object is not NSDate"])
            }
            return nil;
        }
        
        let string = TRCValueTransformerDateISO8601.sharedFormatter.stringFromDate(object as! NSDate)
    
        if !string.isEmpty && error != nil {
            error.memory = NSError(domain: "", code: 0, userInfo: [ NSLocalizedDescriptionKey: "Can't convert NSDate into NSString"])
        }
        
        return string
    }
    
  
}